File: Module 3(Frontend and backend )/Backend/util/otpservice.js
const OTP = require("../models/otp.model");
const sendEmail = require("./sendEmail");

/**
 * Generate, send, and store OTP for an email.
 * @param {string} email
 * @returns {Object} { success: boolean, message: string, otp?: string }
 */
async function generateAndSendOTP(email) {
  try {
    email = email.trim().toLowerCase();

    // Generate OTP
    const otp = Math.floor(100000 + Math.random() * 900000).toString();

    // Send email
    const sent = await sendEmail(email, "Your OTP", `<p>Your OTP is: <strong>${otp}</strong></p>`);
    if (!sent) return { success: false, message: "Failed to send OTP email" };

    // Save or update OTP in DB
    const existingOtp = await OTP.findOne({ email });
    if (existingOtp) {
      existingOtp.otp = otp;
      existingOtp.createdAt = new Date();
      await existingOtp.save();
    } else {
      await new OTP({ email, otp }).save();
    }

    return { success: true, message: "OTP sent successfully", otp };
  } catch (error) {
    console.error("Error in OTP service:", error);
    return { success: false, message: "Internal server error" };
  }
}

/**
 * Verify OTP for an email.
 * @param {string} email
 * @param {string} otp
 * @returns {Object} { success: boolean, message: string }
 */
async function verifyOTP(email, otp) {
  try {
    email = email.trim().toLowerCase();

    const otpEntry = await OTP.findOne({ email });
    if (!otpEntry) return { success: false, message: "OTP not found or expired" };

    if (otpEntry.otp !== otp) return { success: false, message: "Invalid OTP" };

    // OTP verified → delete it
    await OTP.deleteOne({ email });

    return { success: true, message: "OTP verified successfully" };
  } catch (error) {
    console.error("Error verifying OTP:", error);
    return { success: false, message: "Internal server error" };
  }
}

module.exports = { generateAndSendOTP, verifyOTP };

File: Experiment 9 & 10  -Final Project/backend/src/main/java/com/example/backend/service/OtpService.java
package com.example.backend.service;

import com.example.backend.model.Otp;
import com.example.backend.repository.OtpRepository;
import com.example.backend.util.EmailUtil;
import org.springframework.stereotype.Service;

import java.util.Date;
import java.util.Optional;
import java.util.Random;

@Service
public class OtpService {

    private final OtpRepository otpRepository;
    private final EmailUtil emailUtil;

    public OtpService(OtpRepository otpRepository, EmailUtil emailUtil) {
        this.otpRepository = otpRepository;
        this.emailUtil = emailUtil;
    }

    // Generate, send, and store OTP
    public String generateAndSendOTP(String email) {
        try {
            email = email.trim().toLowerCase();

            // Generate 6-digit OTP
            String otp = String.valueOf(100000 + new Random().nextInt(900000));

            // Send email
            boolean sent = emailUtil.sendEmail(email, "Your OTP",
                    "<p>Your OTP is: <strong>" + otp + "</strong></p>");
            if (!sent) return "Failed to send OTP email";

            // Save or update OTP
            Optional<Otp> existingOtp = otpRepository.findByEmail(email);
            if (existingOtp.isPresent()) {
                Otp otpDoc = existingOtp.get();
                otpDoc.setOtp(otp);
                otpDoc.setCreatedAt(new Date());
                otpRepository.save(otpDoc);
            } else {
                Otp newOtp = new Otp(email, otp);
                otpRepository.save(newOtp);
            }

            return "OTP sent successfully";
        } catch (Exception e) {
            e.printStackTrace();
            return "Internal server error";
        }
    }

    // Verify OTP
    public String verifyOTP(String email, String otp) {
        try {
            email = email.trim().toLowerCase();

            Optional<Otp> otpEntry = otpRepository.findByEmail(email);
            if (otpEntry.isEmpty()) return "OTP not found or expired";

            if (!otpEntry.get().getOtp().equals(otp)) return "Invalid OTP";

            otpRepository.deleteByEmail(email);
            return "OTP verified successfully";

        } catch (Exception e) {
            e.printStackTrace();
            return "Internal server error";
        }
    }
}
